File: solutions/javascript/difference-of-squares/1/difference-of-squares.js
//
// This is only a SKELETON file for the 'Difference Of Squares' exercise. It's been provided as a
// convenience to get you started writing code faster.
//


export class Squares {
  constructor(input){
    this.n = input;
  }

  get sumOfSquares(){
    return (this.n * (this.n + 1) * (2 * this.n + 1)) / 6;
  }

  get squareOfSum(){
    const sum = (this.n * (this.n + 1)) / 2;
    return sum * sum;
  }

  get difference(){
    return this.squareOfSum - this.sumOfSquares;
  }
}

// My solution without algorithms
// export class Squares {
//   constructor(input) {
//     this.naturals = input;
//     console.log(this.naturals);
//   }

//   get sumOfSquares() {
//     let sum = 0;
//     for(let i = 0; i <= this.naturals; i++){
//       sum += (i * i);
//     }
//     return sum;
//   }

//   get squareOfSum() {
//     let sum = 0;
//     for(let i = 0; i <= this.naturals; i++){
//       sum += i;
//     }

//     return sum * sum;
//   }

//   get difference() {
//     return Math.abs(this.sumOfSquares - this.squareOfSum);
//   }
// }

File: solutions/cpp/difference-of-squares/1/difference_of_squares.cpp
#include "difference_of_squares.h"

namespace difference_of_squares {

unsigned int square_of_sum(unsigned int num) {
  unsigned int result{0};

  for (unsigned int i{0}; i <= num; ++i) {
    result += i;
  }

  return result * result;
}

unsigned int sum_of_squares(unsigned int num) {
  unsigned int result{0};

  for (unsigned int i{0}; i <= num; ++i) {
    result += i * i;
  }

  return result;
}

unsigned int difference(unsigned int target) {
  return square_of_sum(target) - sum_of_squares(target);
}

} // namespace difference_of_squares

